### PR TITLE
230 implement sign in screen

### PR DESCRIPTION
**Pull request by bogdan-evtushenko**

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -15,6 +15,7 @@
         android:roundIcon="@mipmap/ic_launcher_antivirus_round"
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
+        <activity android:name=".features.auth.SignInActivity"></activity>
```
Comment (yevhenii-kanivets) [inline app/src/main/AndroidManifest.xml:18]: `portrait` and `/>` ?

File: app/src/main/java/com/bogdan/codeforceswatcher/features/auth/SignInActivity.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        )
+        ifPassword.configure(
+                labelTextResId = R.string.password,
+                type = InputField.Type.PASSWORD,
+                action = InputField.Action.Go {
+                    signInWithEmailAndPassword()
+                }
+        )
+
+        // TODO add spinner
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/auth/SignInActivity.kt:37]: Add spinner ;)

File: app/src/main/java/com/bogdan/codeforceswatcher/features/auth/SignInActivity.kt [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    override fun onStop() {
+        super.onStop()
+        store.unsubscribe(this)
+    }
+
+    private fun forgotPassword() {
+        val email = ifEmail.editText.text.trim().toString()
+        AnalyticsController().logEvent(AnalyticsEvents.FORGOT_PASSWORD_PRESSED, mapOf("email" to email))
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/auth/SignInActivity.kt:59]: Well done 👍

File: app/src/main/java/com/bogdan/codeforceswatcher/features/users/ProfileItemEpoxyModel.kt [hunk 4]
```diff
@@ -72,7 +74,8 @@ class ProfileItemEpoxyModel(private val userAccount: UserAccount?) : BaseEpoxyMo
         btnAction.text = context.getString(R.string.login_in_42_seconds)
 
         btnAction.setOnClickListener {
-            store.dispatch(AuthRequests.SignIn())
+            context.startActivity(Intent(context, SignInActivity::class.java))
+            // store.dispatch(AuthRequests.SignIn())
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/users/ProfileItemEpoxyModel.kt:78]: Why commented?

File: app/src/main/java/com/bogdan/codeforceswatcher/util/Prefs.kt [hunk 5]
```diff
@@ -4,11 +4,27 @@ import android.annotation.SuppressLint
...
+    override fun writeUserAccount(userAccount: UserAccount) {
+        val editor = getDefaultPrefs().edit()
+        editor.putString(KEY_USER_ACCOUNT, Gson().toJson(userAccount))
+        editor.apply()
+    }
+
+    override fun readUserAccount(): UserAccount {
+        val defaultPrefs = getDefaultPrefs()
+        return Gson().fromJson<UserAccount>(
+                defaultPrefs.getString(KEY_USER_ACCOUNT, ""),
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/util/Prefs.kt:23]: Won't `fromJson` crash if default `""` passed?

File: common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/auth/AuthRequests.kt [hunk 6]
```diff
@@ -1,29 +1,53 @@
...
-    class SignIn : Request() {
+    class SignIn(
+            private val email: String,
+            private val password: String
+    ) : Request() {
 
         override suspend fun execute() {
-            store.dispatch(Success(UserAccount(null, "token")))
+            val response = backendRepository.signIn(email, password)
+            if (response is Response.Success) {
```
Comment (yevhenii-kanivets) [inline common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/auth/AuthRequests.kt:19]: Hm, `when` should be much more readable here.

File: common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/auth/AuthRequests.kt [hunk 7]
```diff
@@ -1,29 +1,53 @@
...
+                store.dispatch(Success(response.result))
+                settings.writeUserAccount(response.result)
+            } else {
+                with(response as Response.Failure) {
+                    store.dispatch(
+                            Failure(error?.let { Message.Custom(it) } ?: Message.None)
```
Comment (yevhenii-kanivets) [inline common/src/commonMain/kotlin/io/xorum/codeforceswatcher/features/auth/AuthRequests.kt:25]: `error.toMessage()` extension might improve readability.

File: common/src/commonMain/kotlin/io/xorum/codeforceswatcher/network/KtorResponseClient.kt [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+            Response.Failure(null)
+        }
+    }
+
+    suspend fun getError(responseContent: ByteReadChannel): Error {
+        responseContent.readUTF8Line()?.let {
+            return Json(JsonConfiguration.Stable).parse(Error.serializer(), it)
+        }
+        return Error("hello")
```
Comment (yevhenii-kanivets) [inline common/src/commonMain/kotlin/io/xorum/codeforceswatcher/network/KtorResponseClient.kt:38]: `hello` :)

File: common/src/commonMain/kotlin/io/xorum/codeforceswatcher/network/KtorResponseClient.kt [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import io.ktor.utils.io.ByteReadChannel
+import io.ktor.utils.io.readUTF8Line
+import io.xorum.codeforceswatcher.network.responses.Response
+import io.xorum.codeforceswatcher.redux.analyticsController
+import io.xorum.codeforceswatcher.util.stringify
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonConfiguration
+
+class KtorResponseClient {
```
Comment (yevhenii-kanivets) [inline common/src/commonMain/kotlin/io/xorum/codeforceswatcher/network/KtorResponseClient.kt:15]: Very well 👍

Comment (yevhenii-kanivets) [reply common/src/commonMain/kotlin/io/xorum/codeforceswatcher/network/KtorResponseClient.kt:15]: You should also use for `get` and other requests.

**Review by yevhenii-kanivets — CHANGES_REQUESTED** [review]
Crashing on launch 😬 But good job overall 😄

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (bogdan-evtushenko) [reply app/src/main/AndroidManifest.xml:18]: I suppose that CW supports vertical screen.

Comment (yevhenii-kanivets) [reply app/src/main/AndroidManifest.xml:18]: Vertical (portrait) only for simplicity matters.

File: app/src/main/java/com/bogdan/codeforceswatcher/util/Prefs.kt (hunk 5 above)
Comment (bogdan-evtushenko) [reply app/src/main/java/com/bogdan/codeforceswatcher/util/Prefs.kt:23]: Exactly.

File: app/src/main/java/com/bogdan/codeforceswatcher/features/auth/SignInActivity.kt [hunk 10]
```diff
@@ -66,6 +65,10 @@ class SignInActivity : AppCompatActivity(), StoreSubscriber<UsersState> {
     }
 
     override fun onNewState(state: UsersState) {
-        if (state.userAccount != null) finish()
+        when (state.signInStatus) {
+            UsersState.Status.PENDING -> spinner.visibility = View.VISIBLE.also { println("Here here ") }
```
Comment (yevhenii-kanivets) [inline app/src/main/java/com/bogdan/codeforceswatcher/features/auth/SignInActivity.kt:69]: `There there ` 🤦

**Review by yevhenii-kanivets — CHANGES_REQUESTED** [review]
Almost there:
- back icon is missing on "Sign In" screen
- strange color of "Forgot Password" label (should be main blue)

**Review by yevhenii-kanivets — APPROVED** [review]
Ship it!

**Pull request merged**